draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -966,51 +966,85 @@ The hash2base(msg) function maps a string msg of any length into an element of a
 if H=SHA256 and F is a field of characteristic p = 2^255 - 19, then the
 result of reducing H(msg) (a 256-bit integer) modulo p is slightly more likely
-to be a value in \[0, 38\] than a value in \[39, 2^255 - 19). In this example
-the bias is negligible, but in general the bias can be significant.
+to be in \[0, 38\] than if the value were selected uniformly at random.
+In this example the bias is negligible, but in general it can be significant.
 
-To control bias, the input msg should be hashed to an integer comprising more than log2(p) bits.
-In particular, reducing an integer of ceil(log2(p)) + k bits modulo p gives bias
-at most 2^-k, which is a safe choice for a cryptosystem with k-bit security.
-Thus, if H outputs b bits, then H should be evaluated W = ceil((ceil(log2(p)) + k) / b)
-times and the results concatenated to produce a (W * b)-bit integer, which is then reduced modulo p.
-{{hash2base-impl}} details this procedure.
+To control bias, the input msg should be hashed to an integer comprising at
+least ceil(log2(p)) + k bits; reducing this integer modulo p gives bias at
+most 2^-k, which is a safe choice for a cryptosystem with k-bit security.
+To obtain such an integer, hash H with b-bit output should be evaluated W =
+ceil((ceil(log2(p)) + k) / b) times and the results concatenated to produce a
+(W * b)-bit integer. For example, for H=SHA256, k=128-bit security, and p
+a 256-bit prime, W = ceil((256 + 128) / 256) = 2.
+
+{{hash2base-impl}} details the hash2base procedure.
 
 Note that implementors SHOULD NOT use an iterated procedure, i.e., rejection
 sampling. The reason is that these procedures are difficult to implement in constant time,
 and later well-meaning "optimizations" may silently render an implementation
 non-constant-time.
 
-The performance of hash2base may be limited by the length of the input m.
-To prevent this, hash2base first computes m' = H(msg) and then derives the
-required bits from m'. This entails one extra invocation of H, a
-negligible overhead.
+## Performance considerations {#hash2base-perf}
+
+Since hash2base may invoke H multiple times ({{hash2base-sec}}), its
+performance may be limited by the length of the input msg.
+To address this, hash2base first computes H(msg) and then derives the
+required bits from this value via further invocations of H.
+For short messages this entails one extra invocation of H, which is a
+negligible overhead in the context of hashing to elliptic curves.
+
+A related issue is that the random oracle construction of {{rom}} requires
+evaluating two independent hash functions H0 and H1 on msg.
+A standard way to instantiate independent hashes is to append a counter to
+the value being hashed, e.g., H(msg || 0) and H(msg || 1).
+If msg is long, however, this is either inefficient (because it entails hashing
+msg twice) or requires non-black-box use of H (e.g., partial evaluation).
+
+To sidestep both of these issues, hash2base takes a second argument, ctr,
```

Comment:
Regarding the second issue: 
I am not sure whether this proposal maintains or modifies the security properties required by the RO construction.

Reply:
See [my comment](https://github.com/chris-wood/draft-irtf-cfrg-hash-to-curve/pull/116#issuecomment-498057883), below.